Add ERC: Representable Contract State

cfries opened the pull request:
This proposal specifies a minimal interface `IXMLRepresentableState` that lets an EVM contract expose a ***canonical
representation of its state*** via an XML (or JSON) template with machine-readable bindings to its state and view functions.
Off-chain renderers use only `eth_call` at a specific block to materialize that representation.

*Note: The corresponding interface `IJSONRepresentableState` ist part of the ERC, however the binding scheme is currently focussing on XML bindings.*

## Abstract

This ERC introduces `IXMLRepresentableState`, a standard interface and XML binding schema that allows an EVM smart contract to define a static XML template with machine-readable bindings to its state and view functions. Off-chain renderers use this template to build a canonical XML document representing the contract's state at a specific block, without incurring any on-chain gas cost.

A contract that claims to implement `IXMLRepresentableState` MUST be ***XML-complete***: every piece of mutable state that the author considers semantically relevant MUST be represented in the XML via bindings, so that the rendered XML is a complete representation of the contract at a given (chain-id, address, block-number).

## Motivation

Smart contracts can efficiently orchestrate and process the life-cycle of a financial (derivative) product to an extent that they finally represent *the* financial product itself.

At the same time, many applications require a human-readable, machine-parseable representation of that product and its state: valuation oracles need inputs for settlements, smart bonds and other tokenized instruments need legal terms, term sheets or regulatory reports, and on-chain registries, governance modules or vaults benefit from a stable "document view" of their state.

In the traditional off-chain world, such needs are addressed by standards like FpML, the ISDA Common Domain Model, or the ICMA Bond Data Taxonomy. A common pattern is to treat an XML (or similar) document as the definitive source defining the financial product and then generate code to interact with the corresponding data. When a process modifies or updates properties of the product, developers must synchronize the smart contract's internal state with the off-chain XML representation. Today, each project typically invents its own set of view functions and off-chain conventions, so clients need bespoke code to map contract state into XML, JSON, or PDF. This makes interoperability, independent auditing, and reuse of tooling harder.

This ERC inverts that pattern by putting the smart contract in the centre. A contract declares that it implements `IXMLRepresentableState` and defines an interface of representable state. Off-chain renderers can then derive a canonical XML document that reflects the semantically relevant state of the contract at a given (chain-id, address, block-number), using only `eth_call` and a standardized XML binding schema. Rendering happens entirely off-chain and does not change state, so there is no gas cost, yet the resulting XML remains cryptographically anchored to the chain.

Typical use cases include:

- Smart derivative contracts that must present their current state to a valuation oracle or settlement engine.
- Smart bonds and other tokenized financial instruments that must generate legal terms, term sheets, or regulatory and supervisory reports.
- On-chain registries, governance modules, and vaults that want a reproducible, auditable document-style snapshot of their state.

By standardizing the Solidity interface and the XML attribute schema, this ERC allows generic tools to consume any compliant contract without project-specific adapters, and to plug directly into existing XML-based workflows in finance and beyond.

---

## Specification sketch (informal)

Very briefly, beyond the header/abstract/motivation above, the ERC specifies:

- A minimal interface `IXMLRepresentableState` with `xmlTemplate()` returning a UTF‑8 XML document.
- Optional extensions:
    - `IRepresentableStateVersioned` with `stateVersion()` (monotonically increasing version of the representable state).
    - `IRepresentableStateHashed` with `stateHash()` (hash of a canonical state tuple for the representation).
    - Convenience combinations for XML: `IXMLRepresentableStateVersioned`, `IXMLRepresentableStateHashed`, `IXMLRepresentableStateVersionedHashed`.
- A binding schema in the `evmstate` namespace, with attributes like:
    - **single-binding attributes** (no semicolons, exactly one binding):
        - `evmstate:call` / `evmstate:selector` / `evmstate:returns`
        - `evmstate:format` / `evmstate:scale`
        - `evmstate:target` (element text vs. attribute target)
    - **multi-binding attributes** (semicolon-separated lists, interpreted positionally):
        - `evmstate:calls` / `evmstate:selectors` / `evmstate:returnsList`
        - `evmstate:formats` / `evmstate:scales` / `evmstate:targets`

A renderer:

- fetches `xmlTemplate()` at a fixed block tag `B`,
- walks the XML, resolves each `evmstate:*` binding via `eth_call` at `blockTag=B`,
- writes values into element text or attributes (single- or multi-binding mode),
- and fills the context attributes (`evmstate:chain-id`, `evmstate:contract-address`, `evmstate:block-number`).

Snapshot consistency is guaranteed by always using the same block `B` for all calls.

---

## Reference implementation / demos

We have a small set of demo contracts and a Java-based reference renderer (Web3j + DOM XML), including:

- `MinimalInstrument` – owner, notional, currency, maturity, active flag.
- `TestContract` – exercises many type/format combinations (uint/int, hex, decimal+scale, iso8601 date/datetime, bytes as hex/base64, etc.) including a multi-binding example.
- `InterestRateSwapSettleToMarket` – IRS with settle-to-market semantics and `lastSettlement{time,value}`.
- `BondDataTaxonomyDemo` – ICMA BDT-inspired bond (issuer, ISIN, coupon, dates, etc.) rendered via evmstate bindings.

Project Homepage: http://finmath.gitlab.io/representable-contract-state

cfries committed:
Add ERC: Representable Contract State

cfries committed:
Added link to ethereum-magicians.org

cfries committed:
Update ERCS/erc-rcs.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

cfries committed:
Update ERCS/erc-rcs.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

cfries committed:
Update ERCS/erc-rcs.md

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Renamed file/folder erc-8100

cfries committed:
Added array type. Fixed validation errors.

cfries committed:
Fixed links

cfries committed:
Fixed description

cfries committed:
Fixed link

cfries committed:
Removed external link

cfries committed:
Removed external reference

cfries committed:
Fixed typo

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Small rename xmlTemplate -> stateXmlTemplate and added statePart and by-reference uri.

cfries committed:
Fixed links to ERC-6123

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Set version to 0.8.19

cfries committed:
Fixed name of interface in examples

cfries committed:
Extended example with array of struct (settlements)

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

SamWilsn commented on assets/erc-8100/package.json:
Do you need to include this file? I don't think it adds much to your reference implementation.

SamWilsn commented on assets/erc-8100/README.md:
This kinda looks like you're putting proposal content in your assets directory? Don't do that.

cfries replied:
I see the point. I will reduce the content of the README to a simple overview of the content. Also here, the README is still helpful as it is part of the ERC's NPM package.
My next commit will include a cleaned version of the README avoiding overlap with the proposal.

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+At the same time, many applications require a human-readable, machine-parseable representation of that product and its state: valuation oracles need inputs for settlements, smart bonds and other tokenized instruments need legal terms, term sheets or regulatory reports, and on-chain registries, governance modules or vaults benefit from a stable "document view" of their state.
+
+In the traditional off-chain world, such needs are addressed by standards like FpML, the ISDA Common Domain Model, or the ICMA Bond Data Taxonomy. A common pattern is to treat an XML (or similar) document as the definitive source defining the financial product and then generate code to interact with the corresponding data. When a process modifies or updates properties of the product, developers m ...
```
```suggestion
In the traditional off-chain world, such needs are addressed by standards like Financial products Markup Language (FpML), the International Swaps and Derivatives Association (ISDA) Common Domain Model, or the International Capital Market Association (ICMA) Bond Data Taxonomy. A common pattern is to treat an XML (or similar) document as the definitive source defining the financial product and then generate code to interact with the corresponding data. When a process modifies or updates properties of the product, developers must synchronize the smart contract's internal state with the off-chain XML representation. Today, each project typically invents its own set of view functions and off-chain conventions, so clients need bespoke code to map contract state into XML, JSON, or PDF. This makes interoperability, independent auditing, and reuse of tooling harder.
```

Please expand uncommon abbreviations the first time they are used.

SamWilsn replied:
These may be common abbreviations in a specific domain, but not to general Ethereum developers.

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+#### Representable Contract State referenced in Events
+
+Emitting events is an on-chain operation that MUST be paid for by the transaction sender.
```
This is a fact, not a requirement.

```suggestion
Emitting events is an on-chain operation that must be paid for by the transaction sender.
```

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+This ERC deliberately keeps the on-chain type as `uint256` for simplicity and ABI compatibility.
+Tools are free to expose more human-friendly labels (such as `"settlement-context"`) in their
+user interfaces, as long as those labels map unambiguously to the `uint256 partId` used on-chain.
```
Similar comment about rationale

cfries replied:
I have removed that part (was somewhat redundant anyway).

SamWilsn commented on ERCS/erc-8100.md:
````diff
@@ -0,0 +1,1014 @@
+    evmstate:targets="; currency" />
+```
+
+MUST render to something equivalent to:
````
To be honest, the bulk of this proposal is pretty far out of my wheelhouse, but the wording here is a bit suspect: "something equivalent to" is pretty ambiguous, and likely isn't sufficient for a MUST requirement.

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+These attributes are **context bindings**:
+
+- The renderer MUST set `evmstate:chain-id` to the [EIP-155](./eip-155.md) chain ID, as a base-10 string.
```
Add `requires: 155` to your preamble.

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+- **Misrepresentation**: This ERC cannot prevent a malicious contract from claiming to be XML-complete while omitting relevant state from its XML representation. Users and auditors MUST not rely on the XML alone for safety and SHOULD review the contract code and, where necessary, the `stateHash()` encoding if provided.
+
+- **Renderer correctness**: The security and correctness of the final XML representation depend on the correctness of the off-chain renderer. Independent implementations and tests are recommended.
```
Same kind of comment as earlier. Requirements (with UPPERCASE keywords) belong in the specification section, especially if they're necessary to build a secure implementation, but do discuss their implications here.

cfries replied:
I would consider these security aspect as recommendations and have removed the UPPERCASE keywords.

cfries committed:
Update ERCS/erc-8100.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

cfries committed:
Merge branch 'ethereum:master' into feature/resentable-contract-state

cfries committed:
Update ERCS/erc-8100.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

cfries committed:
Added changes required by review

- In erc-8100.md: Added "requires: 155"
- In erc-8100.md: Clarified what a renderer MUST do.
- In erc-8100.md: Have UPPERCASE normative keywords in the specification section only
- Removed xml-bindings.md
- Cleaned README.md

cfries committed:
Added changes required by review

- Removed UPPERCASE keywords from code documentation. Referencing ERC proposal for specification of the normative behaviour.
- Streamlined README.md

cfries committed:
Removed referce to ERC-6123 and kept example more general.

Added informative sequence diagram.

cfries commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+##### Example: [ERC-6123](./eip-6123.md) Settlement Data by Reference
+
+In [ERC-6123](./eip-6123.md), the settlement process is triggered by the `SettlementRequested` event:
```
No, 6123 just served as an example of an ERC that might use. I have removed the reference to 6123 and kept the example illustrative without that reference.

cfries committed:
Removed the somewhat redundant rational.

cfries commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+
+This ERC introduces `IXMLRepresentableState`, a standard interface and XML binding schema that allows an EVM smart contract to define a static XML template with machine-readable bindings to its state and view functions. Off-chain renderers use this template to build a canonical XML document representing the contract's state at a specific block, without incurring any on-chain gas cost.
+
+A contract that claims to implement `IXMLRepresentableState` MUST be *XML-complete*: every piece of mutable state that the author considers semantically relevant MUST be represented in the XML via bindings, so that the rendered XML is a complete representation of the contract at a given (chain-id, address, block-number).
```
I have removed UPPERCASE keywords from the abstract - they are now in the specification.

cfries committed:
Removed left over references to 6123 and fixed rendering of <partId>

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Fixed reference to SVG.

cfries committed:
Shortened description.

cfries committed:
Fixed typo.

cfries committed:
comment on JSON binding

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Update ERCS/erc-8100.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

cfries committed:
Removed requirement from terminology section.

cfries commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1014 @@
+- **XML template**: A well-formed XML document, returned as a UTF-8 string by the contract, that contains placeholder bindings in the `evmstate` namespace.
+- **Binding**: An `evmstate:*` attribute on an XML element that instructs a renderer to fetch a value from the contract (or from chain context) and insert it into the document.
+- **XML representation**: The final XML document obtained by evaluating all bindings of the XML template against the contract at a specific (chain-id, address, block-number) and then removing all `evmstate:*` attributes.
+- **XML-complete contract**: A contract that implements `IXMLRepresentableState` and whose XML representation encodes all semantically relevant mutable state. Informally, if two contracts are bytecode-identical and their XML representations agree at some block, their externally observable behaviour MUST be the same from that block onward.
```
Corrected in commit https://github.com/ethereum/ERCs/pull/1396/changes/152806986af282e9b8d26f08cac4b3a9cf720b25

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Moved larger motivation of "Representable Contract State referenced in Events" to rationale.

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Cnsistently using title case in headings.

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

cfries committed:
Merge branch 'master' into feature/resentable-contract-state

SamWilsn commented on ERCS/erc-8100.md:
````diff
@@ -0,0 +1,1000 @@
+This ERC RECOMMENDS the following minimal scheme for references to partial XML views:
+
+```text
+evmstate://self/part/{partId}[?key={application-specific-key}]
````
Small aside, but might be worth reserving this scheme if it's important.

SamWilsn commented on ERCS/erc-8100.md:
```diff
@@ -0,0 +1,1000 @@
+
+For the array binding profile defined below, array-valued return types MUST NOT be used in multi-binding mode. Implementations that support the array binding profile MUST treat a binding in multi-binding mode whose declared output type is an array as an error. Array handling in this ERC is restricted to single-binding mode on the array container and to single-binding nodes inside the template row ...
+
+#### Example with a Single Binding to the Element's Text:
```
```suggestion
#### Example with a Single Binding to the Element's Text
```

SamWilsn reviewed (approved):
These are pretty minor, please fix them at your leisure.
